web/src/main/java/com/microsoft/applicationinsights/web/internal/correlation/TraceContextCorrelation.java

```diff
@@ -0,0 +1,502 @@
+     * @param request
+     * @return Incoming Traceparent
+     */
+    private static Traceparent extractIncomingTraceparent(HttpServletRequest request) {
+        Traceparent incomingTraceparent = null;
+
+        Enumeration<String> traceparents = request.getHeaders(TRACEPARENT_HEADER_NAME);
+        List<String> traceparentList = getEnumerationAsCollection(traceparents);
+
+        // W3C spec mandates a request should exactly have 1 Traceparent header
+        if (traceparentList.size() != 1) {
+            return null;
+        }
+
+        try {
+            incomingTraceparent = Traceparent.fromString(traceparentList.get(0));
+        } catch (Exception e) {
+            InternalLogger.INSTANCE.error(String.format("Received invalid traceparent header with exception %s, "
+                + "distributed trace might be broken", ExceptionUtils.getStackTrace(e)));
+        }
+        return incomingTraceparent;
+    }
+
+    /**
+     * This method takes incoming traceparent object and creates a new outbound traceparent object
+     * @param incomingTraceparent
+     * @return
+     */
+    private static Traceparent processIncomingTraceparent(Traceparent incomingTraceparent,
+        HttpServletRequest request) {
+
+        Traceparent processedTraceparent = null;
+
+        // If incoming traceparent is null create a new Traceparent
+        if (incomingTraceparent == null) {
+
+            // If BackCompt mode is enabled, read the Request-Id Header
+            if (isW3CBackCompatEnabled) {
+                processedTraceparent = processLegacyCorrelation(request);
+            }
+
+            if (processedTraceparent == null){
+                processedTraceparent = new Traceparent();
+            }
+
+        } else {
+            // create outbound traceparent inheriting traceId, flags from parent.
+            processedTraceparent = new Traceparent(0, incomingTraceparent.getTraceId(), null,
+                incomingTraceparent.getTraceFlags());
+        }
+        return processedTraceparent;
+    }
+
+    /**
+     * This method processes the legacy Request-ID header for backward compatibility.
+     * @param request
+     * @return
+     */
+    private static Traceparent processLegacyCorrelation(HttpServletRequest request) {
+
+        String requestId = request.getHeader(TelemetryCorrelationUtils.CORRELATION_HEADER_NAME);
+
+        if (requestId != null && !requestId.isEmpty()) {
+            String legacyOperationId = TelemetryCorrelationUtils.extractRootId(requestId);
+            RequestTelemetry requestTelemetry = ThreadContext.getRequestTelemetryContext().getHttpRequestTelemetry();
+            requestTelemetry.getContext().getProperties().putIfAbsent("ai_legacyRootID", legacyOperationId);
+            requestTelemetry.getContext().getOperation().setParentId(requestId);
+            return new Traceparent(0, legacyOperationId, null, 0);
+        }
+        return null;
```

Comment:
Whatever goes wrong, start a brand new traceparent. It should not be null

Reply:
```
if (processedTraceparent == null){
                processedTraceparent = new Traceparent();
            }
``` 
Checkout above. If the processLegacyTraceparentReturnsNull, a brand new traceparent is created.

Reply:
But not if request-id is null or empty

Reply:
@lmolkova Well it does, look at this code block carefully.

```
private static Traceparent processLegacyCorrelation(HttpServletRequest request) {

        String requestId = request.getHeader(TelemetryCorrelationUtils.CORRELATION_HEADER_NAME);

        try {
            if (requestId != null && !requestId.isEmpty()) {
                String legacyOperationId = TelemetryCorrelationUtils.extractRootId(requestId);
                RequestTelemetry requestTelemetry = ThreadContext.getRequestTelemetryContext().getHttpRequestTelemetry();
                requestTelemetry.getContext().getProperties().putIfAbsent("ai_legacyRootID", legacyOperationId);
                requestTelemetry.getContext().getOperation().setParentId(requestId);
                return new Traceparent(0, legacyOperationId, null, 0);
            }
        } catch (Exception e) {
            InternalLogger.INSTANCE.error(String.format("unable to create traceparent from legacy request-id header"
                + " %s", ExceptionUtils.getStackTrace(e)));
        }

        return null;
    }
```
if request-id is null or empty it would return null. Then in the function which calls this, it checks if processedTraceparent == null then create a brand new. The case is covered. Anyway, I have added Tests to assert this.